soc/arm/microchip_mec/common/spigen/mec_spi_gen.py

```diff
@@ -0,0 +1,476 @@
+HDR_SPI_CHPHA_MOSI_EDGE_1 = 0x40
+HDR_SPI_CHPHA_MISO_EDGE_1 = 0
+HDR_SPI_CHPHA_MISO_EDGE_2 = 0x80
+
+# User defined constants HDR_SPI_RD_ (0, 1, 2, 3) as per boot rom spec.
+# 1st digit - number of I/O pins used to transmit the opcode
+# 2nd digit - number of I/O pins used to transmit the SPI address
+# 3rd digit - number of pins used to read data from flash
+# 4th digit (if present) - dummy clocks between address and data phase
+HDR_SPI_RD_111 = 0
+HDR_SPI_RD_1118 = 1
+HDR_SPI_RD_1128 = 2
+HDR_SPI_RD_1148 = 3
+
+# Payload parameters
+PLD_LOAD_ADDR = 0xc0000
+PLD_LOAD_ADDR_MEC172X = 0xc0000
+PLD_LOAD_ADDR_MEC152X = 0xe0000
+PLD_ENTRY_ADDR = 0
+PLD_GRANULARITY = 128
+PLD_PAD_SIZE = 128
+PLD_PAD_BYTE = b'\xff'
+
+MCHP_CHAR_P = 0x50
+MCHP_CHAR_H = 0x48
+MCHP_CHAR_C = 0x43
+MCHP_CHAR_M = 0x4D
+
+EC_INFO_BLOCK_SIZE = 128
+ENCR_KEY_HDR_SIZE = 128
+COSIG_SIZE = 96
+TRAILER_SIZE = 160
+TRAILER_PAD_BYTE = b'\xff'
+
+TAG_SPI_LOC = 0
+HDR_SPI_LOC = 0x100
+PLD_SPI_LOC = 0x1000
+
+CRC_TABLE = [0x00, 0x07, 0x0e, 0x09, 0x1c, 0x1b, 0x12, 0x15,
+             0x38, 0x3f, 0x36, 0x31, 0x24, 0x23, 0x2a, 0x2d]
+
+CHIP_DICT = {
+    'mec152x': { 'sram_base': 0xe0000, 'sram_size': 0x40000, 'header_ver': 2 },
+    'mec172x': { 'sram_base': 0xc0000, 'sram_size': 0x68000, 'header_ver': 3 },
+}
+
+CHIP_DEFAULT = 'mec172x'
+SPI_READ_MODE_DEFAULT = 'fast'
+SPI_FREQ_MHZ_DEFAULT = 12
+SPI_MODE_DEFAULT = 0
+SPI_MODE_MIN = 0
+SPI_MODE_MAX = 7
+SPI_DRIVE_STRENGTH_MULT_DEFAULT = "1x"
+SPI_SLEW_RATE_DEFAULT = "slow"
+
+def print_bytes(title, b):
+    """Print bytes or bytearray as hex values"""
+    print("{0} = {{ ".format(title), end='')
+    count = 1
+    for v in b:
+        print("0x{0:02x}, ".format(v), end='')
+        if (count % 8) == 0:
+            print("")
+        count = count + 1
+
+    print("}")
+
+# CRC8-ITU
+#
+def Crc8(crc, data):
+    """Update CRC8 value."""
+    for v in data:
+        crc = ((crc << 4) & 0xff) ^ (CRC_TABLE[(crc >> 4) ^ (v >> 4)])
+        crc = ((crc << 4) & 0xff) ^ (CRC_TABLE[(crc >> 4) ^ (v & 0xf)])
+    return crc ^ 0x55
+
+# MEC172x Boot-ROM TAG is 4 bytes
+# bits[23:0] = bits[31:8] of the Header SPI address
+#  Header location must be a mutliple of 256 bytes
+# bits[31:24] = CRC8-ITU of bits[23:0]
+# return immutable bytes type
+#
+def build_tag(hdr_spi_loc):
+    """Build MEC172x Boot-ROM TAG"""
+    tag = bytearray([(hdr_spi_loc >> 8) & 0xff,
+                   (hdr_spi_loc >> 16) & 0xff,
+                   (hdr_spi_loc >> 24) & 0xff])
+    tag.append(Crc8(0, tag))
+
+    return bytes(tag)
+
+# Build MEC172x Boot-ROM SPI image header
+# Inputs: Header location in SPI Image
+#         Payload(FW binary) location in SPI Image
+#         Payload load address in MEC172x SPI SRAM
+#         Payload entry point address: 0 instructs Boot-ROM to assume
+#           ARM vector table at beginning of payload and reset handler
+#           address is at offset 4 of payload.
+#         Payload length, must be multiple of PLD_GRANULARITY
+#
```

Comment:
How would you feel about moving these comments in the function docstring? That's how it's done traditionally in Python, Doxygen would potentially pick them up as well which is kinda nice (see the "Comment blocks in Python" in https://www.doxygen.nl/manual/docblocks.html).

Applies to this and the two functions above.

Reply:
moved comments to function docstring.